src/widget/raw_event.rs

```diff
@@ -87,3 +82,115 @@ impl ViewContext {
         }
     }
 }
+
+/// Crush all pointer events into a single pointer that counts clicks
+/// and attaches positions to events that don't contain them.
+#[derive(Default)]
+pub struct PointerCrusher {
+    e: MouseEvent,
+    counter: ClickCounter,
+}
+
+impl PointerCrusher {
+    pub fn new() -> Self {
+        PointerCrusher::default()
+    }
+
+    pub fn mods(&mut self, mods: Modifiers) {
+        self.e.mods = mods;
+    }
+
+    pub fn pressed(&mut self, button: PointerButton) -> MouseEvent {
+        self.e.wheel_delta = Vec2::ZERO;
+        self.e.buttons.insert(button);
+        self.e.count = self.counter.count_for_click(self.e.pos);
+        self.e.button = button;
+        self.e.clone()
+    }
+
+    pub fn released(&mut self, button: PointerButton) -> MouseEvent {
+        self.e.wheel_delta = Vec2::ZERO;
+        self.e.buttons.remove(button);
+        self.e.button = button;
+        self.e.clone()
+    }
+
+    pub fn moved(&mut self, pos: Point) -> MouseEvent {
+        self.e.wheel_delta = Vec2::ZERO;
+        self.e.button = PointerButton::None;
+        self.e.pos = pos;
+        self.e.window_pos = pos;
+        self.e.clone()
+    }
+
+    pub fn wheel(&mut self, wheel_delta: Vec2) -> MouseEvent {
+        self.e.wheel_delta = wheel_delta;
+        self.e.button = PointerButton::None;
+        self.e.clone()
+    }
+}
+
+use instant::Instant;
+use std::cell::Cell;
+use std::time::Duration;
+
+// This is the default timing on windows.
+const MULTI_CLICK_INTERVAL: Duration = Duration::from_millis(500);
+// the max distance between two clicks for them to count as a multi-click
+const MULTI_CLICK_MAX_DISTANCE: f64 = 5.0;
+
+/// A small helper for determining the click-count of a mouse-down event.
+///
+/// Click-count is incremented if both the duration and distance between a pair
+/// of clicks are below some threshold.
+#[derive(Debug, Clone)]
+struct ClickCounter {
+    max_interval: Cell<Duration>,
+    max_distance: Cell<f64>,
+    last_click: Cell<Instant>,
+    last_pos: Cell<Point>,
+    click_count: Cell<u8>,
+}
```

Comment:
Out of curiosity why a `Cell` everywhere? Easy access within non-mutable contexts? (it's not needed currently though)

Reply:
I just copied this from Glazier as-is. Seems needless in actual practice (since this will always be tracked at the window level, which is naturally synchronous).